pkg/webhooks/customresourcedefinitions/customresourcedefinitions.go

```diff
@@ -0,0 +1,203 @@
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"slices"
+	"strings"
+
+	hookconfig "github.com/openshift/managed-cluster-validating-webhooks/pkg/config"
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks/utils"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	WebhookName                    string = "customresourcedefinitions-validation"
+	docString                      string = `Managed OpenShift Customers may not change CustomResourceDefinitions in namespaces managed by Red Hat.`
+	privilegedServiceAccountGroups string = `^system:serviceaccounts:(kube.*|openshift.*|default|redhat.*|osde2e-[a-z0-9]{5})`
+)
+
+var (
+	timeout                          int32 = 2
+	allowedUsers                           = []string{"system:admin", "backplane-cluster-admin"}
+	sreAdminGroups                         = []string{"system:serviceaccounts:openshift-backplane-srep"}
+	privilegedServiceAccountGroupsRe       = regexp.MustCompile(privilegedServiceAccountGroups)
+	scope                                  = admissionregv1.NamespacedScope
+	rules                                  = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{
+				admissionregv1.Create,
+				admissionregv1.Update,
+				admissionregv1.Delete,
+			},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{"apiextensions.k8s.io"},
+				APIVersions: []string{"*"},
+				Resources:   []string{"customresourcedefinitions"},
+				Scope:       &scope,
+			},
+		},
+	}
+	log = logf.Log.WithName(WebhookName)
+)
+
+// customresourcedefinitionsruleWebhook validates a customresourcedefinition change
+type customresourcedefinitionsruleWebhook struct {
+	s runtime.Scheme
+}
+
+// We just need a runtime object to get the namespace
+type customResourceDefinition struct {
+	runtime.Object
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *customresourcedefinitionsruleWebhook {
+	scheme := runtime.NewScheme()
+	return &customresourcedefinitionsruleWebhook{
+		s: *scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *customresourcedefinitionsruleWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	//Implement authorized next
+	return s.authorized(request)
+}
+
+func (s *customresourcedefinitionsruleWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	np, err := s.renderCustomResourceDefinition(request)
+	if err != nil {
+		log.Error(err, "Could not render a CustomResourceDefinition from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if hookconfig.IsPrivilegedNamespace(np.GetNamespace()) {
```

Comment:
Does this actually work? CRDs are not namespaced - `Namespaced` here refers to the corresponding CustomResources that are made

Reply:
from https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition CRDs can be Namespaced or Cluster

to get the list of namespaced CRD's on OSD:

oc get crd -o json | jq -r '.items[] | [.metadata.name, .spec.scope] | @tsv' | grep Namespaced

Reply:
> The custom resource created from a CRD object can be either namespaced or cluster-scoped, as specified in the CRD's spec.scope field

but you can verify this in practice by trying to `oc get crd -n ${NS}` - it's not a namespaced object

Reply:
ack, makes sense now. It appears the way to handle this would be:
1. create a list of protected crd names
2. check the incoming request crd .metadata.name in the list of protected crds
3. if true and allowed return admissionctl.Response(Allowed)
4. if true and not allowed return admissionctl.Response(Denied)
5. if false return admissionctl.Response(Allowed)

Reply:
validating by label added in [aa86d0a](https://github.com/openshift/managed-cluster-validating-webhooks/pull/264/commits/aa86d0aa4b230d8687a947917ce79a52b587cfe4) let's use new thread to discuss further changes